### DISCUSSION TITLE
SqlClientBatchingBatcher doesn't work with MicrosoftDataSqlClientDriver

### DISCUSSION THREAD

**Discussion by tompazourek** [nhibernate/nhibernate-core#3258] [Q&A] [answered]
When using the `MicrosoftDataSqlClientDriver`, I'm unable to use `SqlClientBatchingBatcher`.

There's specificaly code that expects that the SqlCommand will be `System.Data.SqlClient.SqlCommand` which doesn't work well with `Microsoft.Data.SqlClient.SqlCommand`.

Is there some way to work around this? Can this be fixed, please?

Specific line: https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/AdoNet/SqlClientBatchingBatcher.cs#L67

See `_currentBatch.Append((System.Data.SqlClient.SqlCommand) Driver.UnwrapDbCommand(batchUpdate));` in the method:

```csharp
public override void AddToBatch(IExpectation expectation)
{
    _totalExpectedRowsAffected += expectation.ExpectedRowCount;
    var batchUpdate = CurrentCommand;
    Driver.AdjustCommand(batchUpdate);
    string lineWithParameters = null;
    var sqlStatementLogger = Factory.Settings.SqlStatementLogger;
    if (sqlStatementLogger.IsDebugEnabled || Log.IsDebugEnabled())
    {
        lineWithParameters = sqlStatementLogger.GetCommandLineWithParameters(batchUpdate);
        var formatStyle = sqlStatementLogger.DetermineActualStyle(FormatStyle.Basic);
        lineWithParameters = formatStyle.Formatter.Format(lineWithParameters);
        _currentBatchCommandsLog.Append("command ")
            .Append(_currentBatch.CountOfCommands)
            .Append(":")
            .AppendLine(lineWithParameters);
    }
    if (Log.IsDebugEnabled())
    {
        Log.Debug("Adding to batch:{0}", lineWithParameters);
    }
    _currentBatch.Append((System.Data.SqlClient.SqlCommand) Driver.UnwrapDbCommand(batchUpdate));

    if (_currentBatch.CountOfCommands >= _batchSize)
    {
        ExecuteBatchWithTiming(batchUpdate);
    }
}
```

**Comment by fredericDelaporte** [comment, marked as answer]
`SqlClientBatchingBatcher` uses `System.Data.SqlClient`, it cannot work with `Microsoft.Data.SqlClient`.

This batcher is implemented through an internal batcher provided by `System.Data.SqlClient`, see [here](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/AdoNet/SqlClientSqlCommandSet.cs).

For `Microsoft.Data.SqlClient`, use the generic batching batcher. (Just leave it un-configured, the driver will use it by default, see [here](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs#L73).)

**Comment by tompazourek** [comment]
I think it's possible to create a new `MicrosoftDataSqlClientBatchingBatcher`. The internals that the existing `SqlClientBatchingBatcher` uses from `System.Data.SqlClient` are the same in `Microsoft.Data.SqlClient`, only from different namespace and assembly.

Why do you think it cannot work? Should I prepare a PR to demonstrate?

> For Microsoft.Data.SqlClient, use the generic batching batcher. (Just leave it un-configured, the driver will use it by default, see [here](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs#L73).)

I'm in the process of migrating a big application from `System.Data.SqlClient` to `Microsoft.Data.SqlClient` due to some security reasons. If I change it to the generic batching batcher, there will be a performance regression, will it not? As I understand it, the goal of `SqlClientBatchingBatcher` is to make the batching faster.

**Reply by bahusoid** [reply]
> the goal of SqlClientBatchingBatcher is to make batching faster.

No. The generic version was written much later to avoid specific implementations (https://github.com/nhibernate/nhibernate-core/pull/1588). I don't expect a big difference but would be interesting to see performance comparison.

> Should I prepare a PR to demonstrate

Sure if it's really faster.

**Reply by fredericDelaporte** [reply]
`SqlClientBatchingBatcher` is from my viewpoint a legacy. It uses reflection to use some hidden and undocumented features of the SqlClient. That is prone to breakage without notice on every update. Of course this is extremely unlikely to occur with the .Net Framework, due to it being almost frozen and anyway, each of its update was done under strong constraints of retro-compatibility. That is far less the case with .Net Core / .Net 5+. 

By the way, only the .Net Framework target of NHibernate does include the System.Data.SqlClient batcher, see the first line of its code file.
https://github.com/nhibernate/nhibernate-core/blob/d419ab3228f0edd3d56105851869f9e56f5d2bdc/src/NHibernate/AdoNet/SqlClientBatchingBatcher.cs#L1
We have not made it available for .Net Core/.Net 5+. See also the SqlClient driver:
https://github.com/nhibernate/nhibernate-core/blob/d419ab3228f0edd3d56105851869f9e56f5d2bdc/src/NHibernate/Driver/SqlClientDriver.cs#L73-L105

So, unless there are performance metrics showing a strong gain in using the "native" batcher, it will be better not to add one for Microsoft.Data.SqlClient.

**Reply by tompazourek** [reply]
If you'd prefer to kill this batcher and not support it in the newer Microsoft.Data.SqlClient, I can understand that.

I'm afraid I'm not able to help with the performance comparison, but I sure would be interested to see some results in case it would be interesting topic for someone to explore.

Thanks for your time and for discussing this with me.

**Reply by roken** [reply]
There is a very material degradation to moving to the generic batching approach because they operate at two different levels.

The `SqlClientBatchingBatcher` is batching down at the RPC level when talking to the database, effectively shoving multiple unique commands into the same network packet.  MS SQL Server is able to process these independent commands efficiently and minimize the communication overhead.  It feels like an oversight to dismiss the idea of using implementation-specific optimizations as legacy.

Conversation within .NET is happening around here:  https://github.com/dotnet/SqlClient/pull/1825

The new batching API is being introduced via DbBatch, SqlBatch in the new SqlClient stack.

The concatenation of the commands, as GenericBatchingBatcher does, with large numbers of parameters hammers the database pretty substantially in our usage.  We're getting 500-1000% worse performance using the Generic Batching under .NET 6.

Our experience:  

Large INSERT statements with 68 values, client and database housed together on the same network.

| Runtime | Batching | Time to process | Comment |
| --- | --- | --- | --- |
| .NET 4.8 | NHibernate Batching = 100 | 40-50ms | This uses SqlClientBatchingBatcher by default |
| .NET 4.8 | No NHibernate Batching | 60-80ms | This demonstrates the claimed 50% perf improvement Ayende made, which is where all this code originates from |
| .NET 6 | NHibernate Batching = 100 | 280-300ms | Slow, and database is visible angry with this too |
| .NET 6 | No NHibernate Batching | 60-80ms | Same as 4.8 with no batching, which makes sense but good to confirm |
| .NET 6 | NHibernate Batching set = 5 | 70-200ms | Still worse than not batching |
| .NET 6 | NHibernate Batching = 2 | 60-80ms | No improvement |

I could not find any batch size where the generic batching provides any improvement, and in fact it only harms performance as batch size goes up.  Our approach here is to simply disable batching in NHibernate.

**Reply by gliljas** [reply]
Work is ongoing to address this in #3461 . I'll add a bit of effort to make it more than a draft.

**Reply by roken** [reply]
Amazing, I missed that!  I started poking at it but will put that on the backburner and look at this PR

**Reply by fredericDelaporte** [reply]
Note that there is a test which is concluding there is a performance improvement with the generic batcher similar to the native batcher with SQL Server, [here](https://github.com/nhibernate/nhibernate-core/blob/d1790e7d3bb408346059e21279635fb7ac1a902d/src/NHibernate.Test/Ado/GenericBatchingBatcherFixture.cs#L75). This test was was done with the SQL Server on the same machine, though.
